framework/Optimizers/Conjugate.py

```diff
@@ -0,0 +1,421 @@
+    varKPlus,modded = self._generateVarsUpdateConstrained(traj, stepSize, gradient, varK)
+    if len(self.optTrajLive) > 1 and self.counter['solutionUpdate'][traj] > 0:
+      removed = self._removeRedundantTraj(traj, varKPlus)
+    else:
+      removed = False
+    if removed:
+      return None
+    #if the new point was modified by the constraint, reset the step size
+    if modded:
+      self.counter['lastStepSize'][traj] = self.paramDict['initialStepSize']
+      self.raiseADebug('Resetting step size for trajectory',traj,'due to hitting constraints')
+    self.queueUpOptPointRuns(traj,varKPlus)
+    self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = varKPlus
+    return varKPlus
+
+  def _createPerturbationPoints(self, traj, optPoint, submit=True):
+    """
+      Creates perturbation points based on a provided NORMALIZED data point
+      @ In, traj, int, integer label for current trajectory
+      @ In, optPoint, dict, current optimal point near which to calculate gradient
+      @ In, submit, bool, optional, if True then submit perturbation points to queue
+      @ Out, points, list(dict), perturbation points
+    """
+    points = []
+    distance = self.paramDict['pertDist'] * max(self.counter['lastStepSize'][traj],self.paramDict['initialStepSize'])
+
+    for i in self.perturbationIndices:
+      direction = self._getPerturbationDirection(i, step = self.counter['varsUpdate'][traj])
+      point = {}
+      index = 0
+      for var in self.getOptVars():
+        size = np.prod(self.variableShapes[var])
+        if size > 1:
+          new = np.zeros(size)
+          for v, origVal in enumerate(optPoint[var]):
+            new[v] = origVal + distance*direction[index]
+            new[v] = self._checkBoundariesAndModify(1.0, 0.0, 1.0, new[v], 0.9999, 0.0001)
+            index += 1
+          point[var] = new
+        else:
+          val = optPoint[var] + distance*direction[index]
+          val = self._checkBoundariesAndModify(1.0, 0.0, 1.0, val, 0.9999, 0.0001)
+          index += 1
+          point[var] = val
+      points.append(point)
+
+      if submit:
+        prefix = self._createEvaluationIdentifier(traj,self.counter['varsUpdate'][traj],i)
+        self.submissionQueue[traj].append({'inputs':point, 'prefix':prefix})
+    return points
+
+  def polakRibierePowellStep(self, traj,alpha, gfkp1=None):
+    """
+
+      @ In, traj, int, the trajectory we are currently considering
+      @ In, alphs, float, step size, or None if no suitable step was found
+      @ In, gfkp1, ndarray, optional, gradient value for xk (xk being the current parameter estimate).
+      @ Out, gNorm, float, norm of the gradient.
+    """
+    xkp1 = self.counter['xk'][traj]
+    yk = gfkp1 - self.counter['oldGradK'][traj]
+    betaK = max(0, np.dot(yk, gfkp1) / self.counter['deltaK'][traj])
+    self.counter['pk'][traj] = -gfkp1 + betaK * self.counter['pk'][traj]
+    gNorm = np.amax(np.abs(gfkp1))
+    return gNorm
```

Comment:
generalize-ble? Base class?

Reply:
No, this method only used inside the conjugate gradient